debug/org.eclipse.debug.core/core/org/eclipse/debug/core/DebugPlugin.java

```diff
@@ -727,22 +734,49 @@ public void stop(BundleContext context) throws Exception {
 	public void start(BundleContext context) throws Exception {
 		super.start(context);
 		new DebugOptions(context);
-		ResourcesPlugin.getWorkspace().addSaveParticipant(getUniqueIdentifier(),
-				new ISaveParticipant() {
-					@Override
-					public void saving(ISaveContext saveContext) throws CoreException {
-						if (fExpressionManager != null) {
-							fExpressionManager.storeWatchExpressions();
+
+		fWorkspaceServiceTracker = new ServiceTracker<>(context, IWorkspace.class, null) {
+
+			@Override
+			public IWorkspace addingService(ServiceReference<IWorkspace> reference) {
+				IWorkspace workspace = super.addingService(reference);
+				try {
+					workspace.addSaveParticipant(getUniqueIdentifier(), new ISaveParticipant() {
+						@Override
+						public void saving(ISaveContext saveContext) throws CoreException {
+							if (fExpressionManager != null) {
+								fExpressionManager.storeWatchExpressions();
+							}
+							Preferences.savePreferences(DebugPlugin.getUniqueIdentifier());
 						}
-						Preferences.savePreferences(DebugPlugin.getUniqueIdentifier());
-					}
-					@Override
-					public void rollback(ISaveContext saveContext) {}
-					@Override
-					public void prepareToSave(ISaveContext saveContext) throws CoreException {}
-					@Override
-					public void doneSaving(ISaveContext saveContext) {}
-				});
+
+						@Override
+						public void rollback(ISaveContext saveContext) {
+						}
+
+						@Override
+						public void prepareToSave(ISaveContext saveContext) throws CoreException {
+						}
+
+						@Override
+						public void doneSaving(ISaveContext saveContext) {
+						}
+					});
+				} catch (CoreException e) {
+					log(e.getStatus());
```

Comment:
In this case you still are tracking the service and will later try to remove the (non registered) participant, you should free the service here and return null to indicate the service is not tracked.

Reply:
Done.